File: 056_MergeIntervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        size = len(intervals)
        if 1 >= size:
            return intervals
        intervals.sort(key = lambda x : x.start)
        res = list()
        for i in xrange(1, size):
            if intervals[i].start <= intervals[i - 1].end:
                intervals[i].start = intervals[i - 1].start
                intervals[i].end = max(intervals[i - 1].end, intervals[i].end)
            else:
                res.append(intervals[i - 1])
        res.append(intervals[-1])
        return res

# This method won't change the input parameters		
def merge(self, intervals):
    out = []
    for i in sorted(intervals, key=lambda i: i.start):
        if out and i.start <= out[-1].end:
            out[-1].end = max(out[-1].end, i.end)
        else:
            out += i,
    return out
            

File: 056_MergeIntervals.cpp
//90.65%
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        int size(intervals.size());
		if(size <= 1)
			return intervals;
		sort(intervals.begin(), intervals.end(), compareInterval);
		vector<Interval> res;
		for(int i = 1; i < size; i++){
			if(intervals[i].start <= intervals[i - 1].end){
				intervals[i].start = intervals[i - 1].start;
				intervals[i].end = max(intervals[i - 1].end, intervals[i].end);
			}else
				res.push_back(intervals[i - 1]);
		}
		res.push_back(intervals[size - 1]);
		return res;
    }
private:
	static bool compareInterval(Interval& v1, Interval& v2){
		return v1.start < v2.start;
	}

};

//In this way, we won't change the input reference parameter
vector<Interval> merge(vector<Interval>& ins) {
    if (ins.empty()) return vector<Interval>{};
    vector<Interval> res;
    sort(ins.begin(), ins.end(), [](Interval a, Interval b){return a.start < b.start;});
    res.push_back(ins[0]);
    for (int i = 1; i < ins.size(); i++) {
        if (res.back().end < ins[i].start) res.push_back(ins[i]);
        else
            res.back().end = max(res.back().end, ins[i].end);
    }
    return res;
}